### PR TITLE
Added prost::Message support

### PR DESCRIPTION
**Pull request by Volkalex28**
I added support for prost::Message. Think it is good feature for this crate :)

### CODE REVIEW

**Comment by ivmarkov** [comment]
Thank you for contributing. :)

Can you run `cargo fmt` and then push? Then I'll comment on the PR shortly.

**Comment by Volkalex28** [comment]
Do I need to fix the merge conflict?

**Comment by ivmarkov** [comment]
I would appreciate that, yes.

File: Cargo.toml [hunk 1]
```diff
@@ -12,12 +12,14 @@ nightly = ["embedded-io?/async", "embedded-svc?/nightly", "embedded-svc?/experim
 notification = ["futures-util", "embassy-sync"]
 edge-net = ["nightly", "dep:edge-net", "embassy-futures", "embedded-io", "log", "serde", "heapless", "postcard"]
 embedded-svc = ["nightly", "dep:embedded-svc", "embassy-futures", "log", "serde", "heapless", "postcard"]
+embedded-svc-prost = ["nightly", "dep:embedded-svc", "embassy-futures", "log", "prost", "heapless", "postcard"]
```
Comment (ivmarkov) [inline Cargo.toml]: The support for `prost` should be implemented for `embedded-svc` (as you did) *but also* for the `edge-net` feature as well.  It is just swapping out the `postcard` serde in the `ws` module with `prost`.

Therefore, best is to just have a `prost` feature, which in fact you can implicitly do by marking the `prost` dependency as optional.

File: src/asynch/pubsub.rs [hunk 2]
```diff
@@ -8,7 +8,7 @@ use embassy_sync::{
 
 use super::{Receiver, Sender};
 
-#[cfg(feature = "embedded-svc")]
+#[cfg(any(feature = "embedded-svc", feature = "embedded-svc-prost"))]
```
Comment (ivmarkov) [inline src/asynch/pubsub.rs]: Revert to the original code.

File: src/asynch/pubsub.rs [hunk 3]
```diff
@@ -97,7 +97,7 @@ where
     }
 }
 
-#[cfg(feature = "embedded-svc")]
+#[cfg(any(feature = "embedded-svc", feature = "embedded-svc-prost"))]
```
Comment (ivmarkov) [inline src/asynch/pubsub.rs]: Ditto: revert to the original code.

File: src/asynch/ws.rs [hunk 4]
```diff
@@ -1,44 +1,90 @@
-#[cfg(any(feature = "edge-net", feature = "embedded-svc"))]
+#[cfg(any(
```
Comment (ivmarkov) [inline src/asynch/ws.rs]: Ditto: revert to the original code.

File: src/asynch/ws.rs [hunk 5]
```diff
@@ -1,44 +1,90 @@
...
+    feature = "embedded-svc",
+    feature = "embedded-svc-prost"
+))]
 pub use error::*;
 
 #[cfg(feature = "edge-net")]
 pub use edge_net_impl::*;
 
-#[cfg(feature = "embedded-svc")]
+#[cfg(any(feature = "embedded-svc", feature = "embedded-svc-prost"))]
```
Comment (ivmarkov) [inline src/asynch/ws.rs]: Ditto.

File: src/asynch/ws.rs [hunk 6]
```diff
@@ -1,44 +1,90 @@
...
 pub use embedded_svc_impl::*;
 
-#[cfg(any(feature = "edge-net", feature = "embedded-svc"))]
+#[cfg(any(
```
Comment (ivmarkov) [inline src/asynch/ws.rs]: Ditto.

File: src/asynch/ws.rs [hunk 7]
```diff
@@ -1,44 +1,90 @@
...
+    feature = "edge-net",
+    feature = "embedded-svc",
+    feature = "embedded-svc-prost"
+))]
 mod error {
     use core::fmt::{self, Debug, Display};
 
+    #[cfg(feature = "embedded-svc-prost")]
```
Comment (ivmarkov) [inline src/asynch/ws.rs]: Now here and further down on all `ProstError` code blocks you should conditionalize by `#[cfg(feature = "prost")]` instead.

File: src/asynch/ws.rs [hunk 8]
```diff
@@ -1,44 +1,90 @@
...
+    impl<E> Display for WsError<E>
     where
-        E: core::fmt::Display,
+        E: Display,
     {
         fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
             match self {
                 Self::IoError(e) => write!(f, "IO Error: {}", e),
                 Self::UnknownFrameError => write!(f, "Unknown Frame Error"),
+                #[cfg(any(feature = "edge-net", feature = "embedded-svc"))]
```
Comment (ivmarkov) [inline src/asynch/ws.rs]: `#[cfg(not(feature = "prost"))]`

File: src/asynch/ws.rs [hunk 9]
```diff
@@ -1,44 +1,90 @@
...
                 Self::PostcardError(e) => write!(f, "Postcard Error: {}", e),
+                #[cfg(feature = "embedded-svc-prost")]
```
Comment (ivmarkov) [inline src/asynch/ws.rs]: `#[cfg(feature = "prost")]`

File: src/asynch/ws.rs [hunk 10]
```diff
@@ -1,44 +1,90 @@
...
+                Self::ProstError(e) => write!(f, "Prost Error {}", e),
             }
         }
     }
 
     #[cfg(feature = "std")]
     impl<E> std::error::Error for WsError<E> where E: Display + Debug {}
 
+    #[cfg(any(feature = "edge-net", feature = "embedded-svc"))]
```
Comment (ivmarkov) [inline src/asynch/ws.rs]: `#[cfg(not(feature = "prost"))]`

File: src/asynch/ws.rs [hunk 11]
```diff
@@ -1,44 +1,90 @@
...
     impl<E> From<postcard::Error> for WsError<E> {
         fn from(e: postcard::Error) -> Self {
             WsError::PostcardError(e)
         }
     }
+
```
Comment (ivmarkov) [inline src/asynch/ws.rs:66]: `#[cfg(feature = "prost")]`

File: src/asynch/ws.rs [hunk 12]
```diff
@@ -1,44 +1,90 @@
...
+    #[cfg(feature = "embedded-svc-prost")]
+    impl<E> From<prost::EncodeError> for WsError<E> {
+        fn from(e: prost::EncodeError) -> Self {
+            WsError::ProstError(ProstError::Encode(e))
+        }
+    }
+
+    #[cfg(feature = "embedded-svc-prost")]
```
Comment (ivmarkov) [inline src/asynch/ws.rs]: Dito.

File: src/asynch/ws.rs [hunk 13]
```diff
@@ -146,23 +192,33 @@ mod edge_net_impl {
     }
 }
 
-#[cfg(feature = "embedded-svc")]
+#[cfg(any(feature = "embedded-svc", feature = "embedded-svc-prost"))]
```
Comment (ivmarkov) [inline src/asynch/ws.rs]: Revert. Also and again: pleae note that ideally you should conditionalize the `edge-net` WS sender/receiver impls with `prost`, not just the `embedded-svc` ones.

File: src/asynch/ws.rs [hunk 14]
```diff
@@ -146,23 +192,33 @@ mod edge_net_impl {
...
     use embedded_svc::ws::{self, FrameType};
 
     use super::*;
 
+    #[cfg(feature = "embedded-svc")]
+    use serde::de::DeserializeOwned as ReceiveData;
+    #[cfg(feature = "embedded-svc")]
+    use serde::Serialize as SendData;
+
+    #[cfg(feature = "embedded-svc-prost")]
```
Comment (ivmarkov) [inline src/asynch/ws.rs]: Ditto and so on for `embedded-svc-prost`.

**Review by ivmarkov — CHANGES_REQUESTED** [review]

**Comment by ivmarkov** [comment]
Changes seem fine!
One final req: can you fix the latest clippy errors (I know these are in the postcard code, not prost so I might be asking too much) but if you have the project checked out and handy...

Then I'll merge.

**Comment by Volkalex28** [comment]
> cargo clippy - done

**Comment by ivmarkov** [comment]
Thanks!

**Pull request merged**